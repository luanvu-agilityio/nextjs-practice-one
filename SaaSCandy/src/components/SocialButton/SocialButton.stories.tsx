import type { Meta, StoryObj } from '@storybook/nextjs';
import SocialButton from './index';

// Mock icons
const GoogleIcon = ({ className }: { className?: string }) => (
  <svg className={className} viewBox='0 0 24 24' fill='currentColor'>
    <path d='M22.56 12.25c0-.78-.07-1.53-.2-2.25H12v4.26h5.92c-.26 1.37-1.04 2.53-2.21 3.31v2.77h3.57c2.08-1.92 3.28-4.74 3.28-8.09z' />
  </svg>
);

const FacebookIcon = ({ className }: { className?: string }) => (
  <svg className={className} viewBox='0 0 24 24' fill='currentColor'>
    <path d='M24 12.073c0-6.627-5.373-12-12-12s-12 5.373-12 12c0 5.99 4.388 10.954 10.125 11.854v-8.385H7.078v-3.47h3.047V9.43c0-3.007 1.792-4.669 4.533-4.669 1.312 0 2.686.235 2.686.235v2.953H15.83c-1.491 0-1.956.925-1.956 1.874v2.25h3.328l-.532 3.47h-2.796v8.385C19.612 23.027 24 18.062 24 12.073z' />
  </svg>
);

const meta: Meta<typeof SocialButton> = {
  title: 'Components/SocialButton',
  component: SocialButton,
  parameters: {
    layout: 'centered',
  },
  decorators: [
    Story => (
      <div style={{ width: '400px' }}>
        <Story />
      </div>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof SocialButton>;

export const Google: Story = {
  args: {
    provider: 'google',
    icon: GoogleIcon,
    children: 'Continue with Google',
  },
};

export const Facebook: Story = {
  args: {
    provider: 'facebook',
    icon: FacebookIcon,
    children: 'Continue with Facebook',
  },
};

export const Disabled: Story = {
  args: {
    provider: 'google',
    icon: GoogleIcon,
    children: 'Continue with Google',
    disabled: true,
  },
};
