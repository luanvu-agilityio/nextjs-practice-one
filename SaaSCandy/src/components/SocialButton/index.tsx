'use client';

import { ComponentType } from 'react';
import { cva, type VariantProps } from 'class-variance-authority';

// Utils
import { cn } from '@/lib/utils';

const socialButtonVariants = cva([
  'flex items-center justify-center gap-3',
  'w-full px-4 py-3 rounded-xl',
  'text-md font-medium transition-colors',
  'border border-gray-300 bg-white text-gray-700',
  'hover:bg-gray-50 hover:border-gray-400',
  'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2',
]);

interface SocialButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof socialButtonVariants> {
  provider: string;
  icon: ComponentType<{ className?: string }>;
  children: React.ReactNode;
}

function SocialButton({
  provider,
  icon: Icon,
  children,
  className,
  ...props
}: Readonly<SocialButtonProps>) {
  return (
    <button
      className={cn(socialButtonVariants(), className)}
      data-provider={provider}
      aria-label={`Sign in with ${provider}`}
      {...props}
    >
      <Icon className='w-5 h-5' />
      {children}
    </button>
  );
}
export default SocialButton;
