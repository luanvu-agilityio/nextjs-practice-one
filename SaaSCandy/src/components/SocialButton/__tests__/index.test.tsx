import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SocialButton from '../index';

// Mock icon component
const MockIcon = ({ className }: { className?: string }) => (
  <svg data-testid='mock-icon' className={className}>
    <path />
  </svg>
);

describe('SocialButton - Snapshot', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <SocialButton provider='google' icon={MockIcon}>
        Continue with Google
      </SocialButton>
    );
    expect(container).toMatchSnapshot();
  });
});

describe('SocialButton - Interactive', () => {
  it('calls onClick when clicked', async () => {
    const user = userEvent.setup();
    const handleClick = jest.fn();

    render(
      <SocialButton provider='google' icon={MockIcon} onClick={handleClick}>
        Continue with Google
      </SocialButton>
    );

    const button = screen.getByRole('button', { name: /sign in with google/i });
    await user.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
