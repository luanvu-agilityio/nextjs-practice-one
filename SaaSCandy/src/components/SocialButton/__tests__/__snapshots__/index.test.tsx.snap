// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SocialButton - Snapshot matches snapshot 1`] = `
<div>
  <button
    aria-label="Sign in with google"
    class="flex items-center justify-center gap-3 w-full px-4 py-3 rounded-xl text-md font-medium transition-colors border border-gray-300 bg-white text-gray-700 hover:bg-gray-50 hover:border-gray-400 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2"
    data-provider="google"
  >
    <svg
      class="w-5 h-5"
      data-testid="mock-icon"
    >
      <path />
    </svg>
    Continue with Google
  </button>
</div>
`;
